core/src/window_service.rs

```diff
@@ -358,28 +447,69 @@ where
             .flat_map_iter(|pkt| pkt.packets.iter().filter_map(handle_packet))
             .unzip()
     });
+
+    let mut retransmit_shreds = Shreds::default();
+    retransmit_shreds.timer = packet_timer;
+    retransmit_shreds.shreds = shreds
+        .iter()
+        .zip(&repair_infos)
+        .filter(|(_, repair_info)| repair_info.is_none())
+        .map(|(shred, _)| shred)
+        .cloned()
+        .collect();
+    // TODO propagate timer from incoming packets
+    retransmit_shreds.timer.mark_outgoing_start();
+
     // Exclude repair packets from retransmit.
-    let _ = retransmit_sender.send(
-        shreds
-            .iter()
-            .zip(&repair_infos)
-            .filter(|(_, repair_info)| repair_info.is_none())
-            .map(|(shred, _)| shred)
-            .cloned()
-            .collect(),
-    );
+    let _ = retransmit_sender.send(retransmit_shreds);
+
     stats.num_repairs += repair_infos.iter().filter(|r| r.is_some()).count();
     stats.num_shreds += shreds.len();
     for shred in &shreds {
         *stats.slots.entry(shred.slot()).or_default() += 1;
     }
-    insert_shred_sender.send((shreds, repair_infos))?;
+    let mut insert_shreds = Shreds::default();
+    insert_shreds.timer = packet_timer;
+    insert_shreds.timer.mark_outgoing_start();
+    insert_shreds.shreds = shreds;
+    insert_shred_sender.send((insert_shreds, repair_infos))?;
 
     stats.num_packets += packets.iter().map(|pkt| pkt.packets.len()).sum::<usize>();
     for packet in packets.iter().flat_map(|pkt| pkt.packets.iter()) {
         *stats.addrs.entry(packet.meta.addr()).or_default() += 1;
     }
+
+    let coalesced_batches: usize = packets
+        .iter()
+        .map(|pkts| pkts.timer.get_num_coalesced())
+        .sum();
     stats.elapsed += now.elapsed();
+    stats.num_packet_batches += packets.len();
+    stats.num_coalesced_batches += coalesced_batches;
+    stats.elapsed_first_recv_to_start +=
+        now - packets.first().unwrap().timer.get_incoming_start().unwrap();
+    stats.elapsed_last_recv_to_start +=
+        now - packets.last().unwrap().timer.get_incoming_end().unwrap();
```

Comment:
Will this aggregate information be useful without know how many `packets` we've processed? I see above we increment `num_packet_batches` by `packets.len()`, but we only add to `elapsed_first_recv_to_start` once per `packets`

Reply:
removed these stats, they're captured by the histograms